### ISSUE TITLE
Weird transcriptions from a fine-tuned model

### ISSUE THREAD

**Issue by BenoitWang** [NVIDIA/NeMo#3396]
Hi, 

I fine-tuned a pre-trained French ASR model "stt_fr_conformer_ctc_large" with my own data. And I replaced the tokenizer with my own by `model.change_vocabulary(new_tokenizer_dir=my_own_tokenizer, new_tokenizer_type="bpe")`.

The training was smooth and I got 17% valid WER which is rather good for our dataset, but when I did `transcription=asr_model.transcribe(["example.wav"])` for inference, I got weird transcriptions as below : 
![image](https://user-images.githubusercontent.com/41187612/148400356-a24dfafb-369e-4c53-aa40-3a9ac349cf13.png)
It seems that something changed the alphabet, and here is another example:
![image](https://user-images.githubusercontent.com/41187612/148400928-3c1e21b6-68a7-4781-a226-e47d59afca8c.png)
I looked back into the training log and I found the same thing : 
![image](https://user-images.githubusercontent.com/41187612/148401547-937142cb-f1de-4ab4-8a90-83fa430d37d2.png)

How can I get good transcriptions please ? And is my 17% WER correct ?
Thanks in advance.

**Comment by titu1994** [comment]
You should note you r validation wer, not your training loss or training wer. After changing vocabulary, do you use the same model to transcribe ? Ie I see model change_vocab but asr_model.transcribe(). 

In the models config you should be able to print your models vocabulary from the decoder as well as the Tokenizers vocabulary. 

You can use the speech to text eval script https://github.com/NVIDIA/NeMo/blob/main/examples/asr/speech_to_text_eval.py to get the raw predictions + the final offline wer on a dataset

**Comment by BenoitWang** [comment]
Hi, thanks for the reply. The problem I want to spot is rather during the fine-tuning process than the inference process. In fact in the third graph the reference should be "ouais" instead of "p en donc". I don't know why changing the vocabulary changed the reference too.

**Comment by titu1994** [comment]
Ah ok I just saw that now. The ground truth undergoes the same Tokenizers - Detokenizer processing as the model - this is to ensure exactly the case you have faced can be easily diagnosed. 

Most possibly your tokenizer is incorrect - it is not processing the original text correctly and using incorrect subword token to process it so the output is incorrect. 

You need to check how the tokenizer was built and how it preprocessed the dataset. Did you follow the tutorial when building the tokenizer ?

**Comment by titu1994** [comment]
Also in such a case wer of 17% is incorrect. You'll need to first have correct ground truth after tokenizer - Detokenizer phase to properly train the model.

**Comment by BenoitWang** [comment]
Hi,

The problem was that `model.change_vocabulary()` and  `model.setup_training_data()` were in wrong order. I should just first update the datasets and then the tokenizer, and it works fine. 

Just another question please, I found that fine-tune on 16k data is much faster then on 8k data which is fairly logic, but I got better WER with 8k data during inference directly from the pretrained model (without fine-tuning), which I do not really understand. Any ideas please?

**Comment by titu1994** [comment]
That would be incorrect. Your tokenizer needs to be updated first - it is what's used to tokenize your dataset text into tokens. If you put it after dataset preparation then your tokenizer is incorrect and unused.

**Comment by titu1994** [comment]
You need to change sample rate of model to 8k, but it will be incompatible with pretrained weights. Right now Nemo is dynamically upsampling 8 to 16 khz to preserve compatibility so it's slower c

**Comment by BenoitWang** [comment]
Oh sorry for the wrong message. Yes as you said I had first updated the tokenizer and then the datasets, and it worked.
Thanks
